### ISSUE TITLE
Conversion strategy

### ISSUE THREAD

**Issue by STRML** [brigand/babel-plugin-flow-react-proptypes#9]
I'm working to convert [react-grid-layout](https://github.com/STRML/react-grid-layout) to use this plugin.

I'm running into an issue where the plugin crashes on the IntersectionTypeAnnotation [here](https://github.com/STRML/react-grid-layout/blob/master/lib/utils.js#L3). That's fine - and I'll submit a PR soon to try to address it or ignore it (perhaps we need suppression pragmas much like Flow itself). Also will submit one to add some context to the error.

My issue is - why is it even processing this node at all? I don't like the idea of matching type names like `IFooProps` - why not just convert the node that we know has been attached to a React Component's `props`?

**Comment by brigand** [comment]
I don't like it either. It's a lazy implementation.

Is this what you want to match?

```js
type X = {};

class Foo extends React.Component {
  props: X;
}
```

I'm not sure how this'd be implemented, so I'm glad you volunteered :-)

**Comment by STRML** [comment]
Ahahaha. Well, time to spend some quality time in astexplorer and figure it out.

**Comment by STRML** [comment]
Stalled right now as astexplorer broke our usage (https://github.com/fkling/astexplorer/pull/112) - when I have time I'll work on a PR.

**Comment by JakeCoxon** [comment]
I'm not a big fan of matching on the name of the type either.

What if you denote a special generic type to use for the plugin.

```
type X = {};

class Foo extends React.Component {
  props: ReactProps<X>;
}
```

Then anything with ReactProps<Y> will be transformed

**Comment by brigand** [comment]
@JakeCoxon how would that work with people using flow, wouldn't it error saying ReactProps doesn't exist?

**Comment by STRML** [comment]
Could use a passthrough type that we import from this module. A bit hacky but should work:

```js
export type ReactProps<T> = T;
```

Not sure if this is the best option though.

**Comment by brigand** [comment]
So far, I think your original idea of checking `props: SomeType` in the class is best.

**Comment by JakeCoxon** [comment]
Sure the passthrough would work, it is a pain to have to import it, but if you're using flow then you don't need this plugin right? For production you can just replace ReactProps<T> with T.

I definitely think semantics based on the identifier name not good enough.

Otherwise you will have to check each class to see if it extends React.Component? But what if you import it under a different name? What if you extend a different class. How will pure function components work?

I would really like this to work 

```const comp = (props: SomeType) => <div />```

But I just don't see how you can know when to convert that into React's prop types, without giving the plugin some extra information.

**Comment by STRML** [comment]
That's a good point re: functional components.

The pain of importing `ReactProps` could be lessened somewhat by encouraging that you install a global interface. It becomes a bit like `__()` in i18n-land.

Otherwise, we could duck-type any class with a `render()` method or function that returns a React component. But that could be pretty difficult to do right, especially if the method sometimes returns `null`.

**Comment by STRML** [comment]
If it's hard to identify a ReactClass directly, perhaps we could trace backwards? If we see any React.createElement() calls or JSX custom components in the code, and they're not in `node_modules`, we could then mark the import as a ReactClass or functional component & do the transformation. I'm not familiar enough with Babel to know if this would be possible at all, or at least without requiring multiple passes.

**Comment by brigand** [comment]
I'm not sure if checking if it uses jsx/createElement is much better than the current method. And it'd have the problem with e.g. `import {createElement} from 'react';`. 

We recently got support for some function components (https://github.com/brigand/babel-plugin-flow-react-proptypes/pull/27) but it has the same "ends with Props" hack.

**Comment by dozoisch** [comment]
@brigand and that end with Props hack seems to break when you have a constructor in a class :)

Currently It prints the propTypes twice. Once inside the class declaration, and once outside and it seems to break in react-native. I'm investigating a fix right now.

repro code:

```
var React = require('react');

type Props = {
  an_optional_string?: string,
  a_number: number,
};

export default class Foo extends React.Component<any, Props, void> {
  props: Props;
  _nonEmptyCtor: number;

  constructor(props: Props) {
    super(props);
    this._nonEmptyCtor = 7;
  }
}
```

**Comment by STRML** [comment]
Appears this is fixed as of 0.9.0?

**Comment by STRML** [comment]
We still have some issues with extended components, like so:

```
export default class SomeComponent<DefaultProps, Props, State, Context>
               extends Component<DefaultProps, Props, State> {
  // Type parameters break unless you use them somehow
  // Flow bug - see flow/lib/react.js where they realize this themselves
  static defaultProps: $Abstract<DefaultProps>;
  props: Props;
  //...
```

Throws:

```
Module build failed: Error: /project/ui/views/SomeComponent.jsx: Did not find type annotation for SomeComponent
```

A suppression pragma would be really nice.

**Comment by brigand** [comment]
@STRML Good idea! Published in 0.13.0. https://github.com/brigand/babel-plugin-flow-react-proptypes#suppresion

**Comment by STRML** [comment]
Cool, thanks! Although you might want to fix the typo: (`pluigin`) 

```
'no babel-pluign-flow-react-proptypes'
```

Also, the README typos `Suppression`.

**Comment by brigand** [comment]
Whoops. Published as 0.13.1.

**Comment by mhaas** [comment]
Is there any remaining issue here? Even if I wanted to fix this, I don't know what's broken ;)
